### PR TITLE
8268169: The system lookup can not find stdio functions such as printf on Windows 10

### PR DESCRIPTION
**Pull request by JornVernee**
Hi,

This patch adds a way to look up stdio functions such as printf on Windows 10 through the system lookup. A fallback library is used which forces the generation of these inline functions, and exposes them through a table of function pointers.

Thanks,
Jorn

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268169](https://bugs.openjdk.java.net/browse/JDK-8268169): The system lookup can not find stdio functions such as printf on Windows 10

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer) ⚠️ Review applies to daa560e79d2e70539b1f4e807656902d67dd2e2b

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/547/head:pull/547` \
`$ git checkout pull/547`

Update a local copy of the PR: \
`$ git checkout pull/547` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/547/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 547`

View PR using the GUI difftool: \
`$ git pr show -t 547`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/547.diff">https://git.openjdk.java.net/panama-foreign/pull/547.diff</a>

</details>

### CODE REVIEW

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SystemLookup.java [hunk 1]
```diff
@@ -45,29 +53,129 @@ private SystemLookup() { }
...
+        wscanf,
+        wscanf_s,
+
+        // time
+        gmtime
+        ;
+
+        private static final Map<String, WindowsFallbackSymbols> BY_NAME;
+
+        static {
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SystemLookup.java:174]: Wouldn't Enum.valueOf achieve the same result here?

File: src/jdk.incubator.foreign/windows/native/libWinFallbackLookup/WinFallbackLookup.c [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    &vwscanf,
+    &vwscanf_s,
+    &wprintf,
+    &wprintf_s,
+    &wscanf,
+    &wscanf_s,
+
+    // time.h
+    &gmtime
+};
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/windows/native/libWinFallbackLookup/WinFallbackLookup.c:86]: Missing newline here

File: test/jdk/java/foreign/StdLibTest.java [hunk 3]
```diff
@@ -155,23 +155,7 @@ void test_vprintf(List<PrintfArg> args) throws Throwable {
...
-            SymbolLookup fallbackLookup = name -> switch (name) {
-                    case "printf" -> Optional.of(MemoryAccess.getAddressAtIndex(funcs, 0));
-                    case "vprintf" -> Optional.of(MemoryAccess.getAddressAtIndex(funcs, 1));
-                    case "gmtime" -> Optional.of(MemoryAccess.getAddressAtIndex(funcs, 2));
-                    default -> Optional.empty();
-                };
-
-            LOOKUP = name -> CLinker.systemLookup().lookup(name).or(() -> fallbackLookup.lookup(name));
-        }
+        final static SymbolLookup LOOKUP = CLinker.systemLookup();
```
Comment (mcimadamore) [inline test/jdk/java/foreign/StdLibTest.java:158]: Nice simplification!

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SystemLookup.java (hunk 1 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SystemLookup.java:174]: Right, I'm so used to the static map idiom for when the key is something else than the name that I didn't really think about using Enum.valueOf.

I'll switch to using valueOf.

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**